blaze/src/lib.rs

```diff
@@ -122,47 +117,99 @@ pub extern "system" fn Java_org_apache_spark_sql_blaze_JniBridge_callNative(
-            } else {
-                JObject::null()
+                // throw jvm runtime exception
+                let cause = if env.exception_check()? {
+                    let throwable = env.exception_occurred()?.into();
+                    env.exception_clear()?;
+                    throwable
+                } else {
+                    JObject::null()
+                };
+                let msg = env.new_string(&panic_message)?;
+                let _throw = jni_bridge_call_static_method_no_check_java_exception!(
+                    env,
+                    JniBridge.raiseThrowable,
+                    jni_bridge_new_object!(env, JavaRuntimeException, msg, cause)?
+                );
+                Ok(())
             };
-            let msg = env.new_string(&panic_message)?;
-            let _throw = jni_bridge_call_static_method_no_check_java_exception!(
-                env,
-                JniBridge.raiseThrowable,
-                jni_bridge_new_object!(env, JavaRuntimeException, msg, cause)?
-            );
-            Ok(())
-        };
-        recover().unwrap_or_else(|err: Box<dyn Error>| {
-            error!("Error recovering from panic, cannot resume: {:?}", err);
-            std::process::abort();
-        });
-    }
+            recover().unwrap_or_else(|err: Box<dyn Error>| {
+                error!("Error recovering from panic, cannot resume: {:?}", err);
+                std::process::abort();
+            });
+            -1
+        }
+        Ok(ptr) => ptr,
+    };
 
     let time_cost_sec = Instant::now().duration_since(start_time).as_secs_f64();
     info!("blaze_call_native() time cost: {} sec", time_cost_sec);
+    result
+}
+
+#[allow(non_snake_case)]
+#[no_mangle]
+pub extern "system" fn Java_org_apache_spark_sql_blaze_JniBridge_loadNext(
+    _: JNIEnv,
+    _: JClass,
+    iter_ptr: i64,
+    schema_ptr: i64,
+    array_ptr: i64,
+) -> i64 {
+    unsafe {
+        let sync_iter = &mut *(iter_ptr as *mut SyncBatchIterator);
+        match sync_iter.next() {
+            Some(Ok(batch)) => {
+                let array: StructArray = batch.into();
+                let out_schema = schema_ptr as *mut FFI_ArrowSchema;
+                let out_array = array_ptr as *mut FFI_ArrowArray;
+                export_array_into_raw(Arc::new(array), out_array, out_schema).unwrap();
+                1
+            }
+            _ => -1,
+        }
+    }
+}
+
+struct SyncBatchIterator {
+    inner: SendableRecordBatchStream,
+    rt: Runtime,
+}
+
+impl SyncBatchIterator {
+    fn new(inner: SendableRecordBatchStream) -> Self {
+        Self {
+            inner,
+            rt: tokio::runtime::Builder::new_current_thread()
+                .build()
+                .unwrap(),
+        }
+    }
+}
+
+impl Iterator for SyncBatchIterator {
+    type Item = ArrowResult<RecordBatch>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.rt.block_on(async { self.inner.next().await })
```

Comment:
problematic here:

```
Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.Cannot start a runtime from within a runtime. This happens because a function (like `block_on`)
```